MAINT Parameters validation for sklearn.random_projection.johnson_lindenstrauss_min_dim

DeaMariaLeon opened the pull request:
Towards #24862

#### Any other comments?

In the docstrings of the function,  the `n_samples` parameter is described as int, or array-like of int. But the examples (in the docstring) are done with `n_samples` as float and as an array of floats. Also in the docstring, the `eps` parameter is not given as array but a list. Testing the docstrings fails then. 

Questions:
1) Is there a way to set the dtype to an “array-like” parameter in `@validate_params`?
Changing the `n_samples` from float to int in the docstring first 2 examples, the docstring test passes for this issue. But when `n_samples` is an array of floats like in the 3rd example, the dtype is not checked. 

2) Should I change the `eps` to array in the examples too?

I found this because I thought there was a little grammar issue. There wasn't, but it made me test the docstrings.

DeaMariaLeon committed:
Parameters validation

DeaMariaLeon committed:
Merging
Merge remote-tracking branch 'upstream/main' into make-auto

glemaitre commented on sklearn/random_projection.py:
```diff
@@ -51,11 +51,17 @@
+@validate_params(
+    {
+        "n_samples": ["array-like", Interval(Integral, 1, None, closed="left")],
+        "eps": [np.ndarray, Interval(Real, 0, 1, closed="neither")],
```
```suggestion
        "eps": ["array-like", Interval(Real, 0, 1, closed="neither")],
```

glemaitre replied:
We can have an array-like since we call `asarray` on it.

glemaitre replied:
We can update the docstring accordingly.

glemaitre reviewed (commented):
Otherwise LGTM.

glemaitre commented:
> Is there a way to set the dtype to an “array-like” parameter in @validate_params?

Nop. This would be an additional check in the function. Here, I would not do anything in the PR because `n_samples=[1000., 1234.]` would be failing if we check for the dtype because they are floating numbers.

In this PR, I would keep as-is. We can check in a later PR if we want to be less permissive.

> Should I change the eps to array in the examples too?

We can keep the example as-is.

DeaMariaLeon committed:
corrected eps to array-like

DeaMariaLeon committed:
merging

glemaitre commented on sklearn/tests/test_public_functions.py:
```diff
@@ -120,6 +120,7 @@ def _check_function_param_validation(
     "sklearn.metrics.roc_curve",
     "sklearn.metrics.zero_one_loss",
     "sklearn.model_selection.train_test_split",
+    "sklearn.random_projection.johnson_lindenstrauss_min_dim"
```
```suggestion
    "sklearn.random_projection.johnson_lindenstrauss_min_dim",
```

DeaMariaLeon committed:
added forgotten comma

DeaMariaLeon commented:
Suggested changes done.
Is the docstring test failing or am I missing something?

jeremiedbb commented:
The example calls ``johnson_lindenstrauss_min_dim(1e6, eps=0.5)``. `1e6` is actually a float. I think we need to use an Interval of ``Real`` for the constraint to not break backward compat. In addition, looking at the code, having a float is really not an issue since we don't rely on it being an integer at all.

DeaMariaLeon committed:
changed n_samples to float

DeaMariaLeon committed:
merging
Merge remote-tracking branch 'upstream/main' into make-auto

DeaMariaLeon commented:
Changed `n_samples`'s Interval from `Integer` to `Real` so the docstrig examples test passes.

glemaitre reviewed (approved):
LGTM. Thanks @DeaMariaLeon

jeremiedbb commented on sklearn/random_projection.py:
```diff
@@ -120,15 +126,6 @@ def johnson_lindenstrauss_min_dim(n_samples, *, eps=0.1):
     eps = np.asarray(eps)
     n_samples = np.asarray(n_samples)
 
-    if np.any(eps <= 0.0) or np.any(eps >= 1):
-        raise ValueError("The JL bound is defined for eps in ]0, 1[, got %r" % eps)
-
-    if np.any(n_samples) <= 0:
-        raise ValueError(
-            "The JL bound is defined for n_samples greater than zero, got %r"
-            % n_samples
-        )
```
We need to keep this check for the array-like case.

jeremiedbb commented on sklearn/tests/test_random_projection.py:
```diff
@@ -64,14 +64,6 @@ def densify(matrix):
 ###############################################################################
 
 
-@pytest.mark.parametrize(
-    "n_samples, eps", [(100, 1.1), (100, 0.0), (100, -0.1), (0, 0.5)]
-)
-def test_invalid_jl_domain(n_samples, eps):
-    with pytest.raises(ValueError):
-        johnson_lindenstrauss_min_dim(n_samples, eps=eps)
-
-
```
We also to need to keep this, but we should change the parametrization to use lists.

DeaMariaLeon committed:
Kept and corrected array-like test, added list test

DeaMariaLeon committed:
merging
Merge remote-tracking branch 'upstream/main' into make-auto

DeaMariaLeon commented:
Added back tests to random_projection.py, but there was bug: `np.any(n_samples)` is always `True` when given an array. Unless it's an array of zeros, if I'm not mistaken. 

Also added a test with lists to test_random_projection.py.

jeremiedbb reviewed (approved):
LGTM. Thanks @DeaMariaLeon

jeremiedbb commented:
> but there was bug

good catch !
